File: emarsys-sdk/src/main/java/com/emarsys/config/ConfigProxy.java
package com.emarsys.config;

import com.emarsys.core.Callable;
import com.emarsys.core.RunnerProxy;
import com.emarsys.core.api.result.CompletionListener;
import com.emarsys.core.util.Assert;

public class ConfigProxy implements ConfigApi {

    private final ConfigInternal configInternal;
    private final RunnerProxy runnerProxy;

    public ConfigProxy(RunnerProxy runnerProxy, ConfigInternal configInternal) {
        Assert.notNull(runnerProxy, "RunnerProxy must not be null!");
        Assert.notNull(configInternal, "ConfigInternal must not be null!");
        this.configInternal = configInternal;
        this.runnerProxy = runnerProxy;
    }

    @Override
    public int getContactFieldId() {
        return runnerProxy.logException(new Callable<Integer>() {
            @Override
            public Integer call() {
                return configInternal.getContactFieldId();
            }
        });
    }

    @Override
    public void changeApplicationCode(final String applicationCode, final CompletionListener completionListener) {
        runnerProxy.logException(new Runnable() {
            @Override
            public void run() {
                configInternal.changeApplicationCode(applicationCode, completionListener);
            }
        });
    }

    @Override
    public String getApplicationCode() {
        return runnerProxy.logException(new Callable<String>() {
            @Override
            public String call() {
                return configInternal.getApplicationCode();
            }
        });
    }

    @Override
    public void changeMerchantId(final String merchantId) {
        runnerProxy.logException(new Runnable() {
            @Override
            public void run() {
                configInternal.changeMerchantId(merchantId);
            }
        });
    }
}

File: emarsys-sdk/src/main/java/com/emarsys/config/ConfigProxy.kt
package com.emarsys.config

import com.emarsys.core.Callable
import com.emarsys.core.RunnerProxy
import com.emarsys.core.api.notification.NotificationSettings
import com.emarsys.core.api.result.CompletionListener

class ConfigProxy(private val runnerProxy: RunnerProxy, private val configInternal: ConfigInternal) : ConfigApi {
    override val contactFieldId: Int
        get() = runnerProxy.logException(Callable { configInternal.contactFieldId })

    override val applicationCode: String?
        get() = runnerProxy.logException(Callable<String> { configInternal.applicationCode })

    override val merchantId: String?
        get() = runnerProxy.logException(Callable<String> { configInternal.merchantId })
    override val hardwareId: String
        get() = runnerProxy.logException(Callable<String> { configInternal.hardwareId })
    override val language: String
        get() = runnerProxy.logException(Callable<String> { configInternal.language })
    override val notificationSettings: NotificationSettings
        get() = runnerProxy.logException(Callable<NotificationSettings> { configInternal.notificationSettings })

    override fun changeApplicationCode(applicationCode: String?) {
        changeApplicationCode(applicationCode, contactFieldId)
    }

    override fun changeApplicationCode(applicationCode: String?, completionListener: CompletionListener?) {
        changeApplicationCode(applicationCode, contactFieldId, completionListener)
    }

    override fun changeApplicationCode(applicationCode: String?, completionListener: (throwable: Throwable?) -> Unit) {
        changeApplicationCode(applicationCode, CompletionListener { completionListener(it) })
    }

    override fun changeApplicationCode(applicationCode: String?, contactFieldId: Int) {
        changeApplicationCode(applicationCode, contactFieldId, null)
    }

    override fun changeApplicationCode(applicationCode: String?, contactFieldId: Int, completionListener: CompletionListener?) {
        runnerProxy.logException { configInternal.changeApplicationCode(applicationCode, contactFieldId, completionListener) }
    }

    override fun changeApplicationCode(applicationCode: String?, contactFieldId: Int, completionListener: (throwable: Throwable?) -> Unit) {
        changeApplicationCode(applicationCode, contactFieldId, CompletionListener { completionListener(it) })
    }

    override fun changeMerchantId(merchantId: String?) {
        runnerProxy.logException { configInternal.changeMerchantId(merchantId) }
    }
}
